File: source/74.搜索二维矩阵.cpp
/*
 * @lc app=leetcode.cn id=74 lang=cpp
 *
 * [74] 搜索二维矩阵
 */
#include "cpp_includes.h"
// @lc code=start
class Solution {
 public:
  bool searchMatrix(vector<vector<int>>& matrix, int target) {
    if (matrix.empty() || matrix[0].empty()) return false;

    int m = matrix.size();
    int n = matrix[0].size();

    int left = 0;
    int right = n * m - 1;

    while (left <= right) {
      cout << left << ',' << right << endl;
      int mid = left + (right - left) / 2;
      int val = matrix[mid / n][mid % n];
      if (val == target) {
        return true;
      } else if (val < target) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return false;
  }
};
// @lc code=end

void main() {
  vector<vector<int>> matrix = {
      {1, 3, 5, 7}, {10, 11, 16, 20}, {23, 30, 34, 50}};
  cout << (new Solution)->searchMatrix(matrix, 3);
}

File: source/74.搜索二维矩阵.py
#
# @lc app=leetcode.cn id=74 lang=python
#
# [74] 搜索二维矩阵
#

# @lc code=start
class Solution(object):
    # def searchMatrix(self, matrix, target):
    #     """
    #     :type matrix: List[List[int]]
    #     :type target: int
    #     :rtype: bool
    #     """
    #     M, N = len(matrix), len(matrix[0])
    #     left, right = 0, M * N - 1
    #     while left <= right:
    #         mid = left + (right - left) // 2
    #         cur = matrix[mid // N][mid % N]
    #         if cur == target:
    #             return True
    #         elif cur < target:
    #             left = mid + 1
    #         else:
    #             right = mid - 1
    #     return False


    def searchMatrix(self, matrix, target):
        M, N = len(matrix), len(matrix[0])
        for i in range(M):
            for j in range(N):
                if matrix[i][j] == target:
                    return True
        return False

# @lc code=end

